### ISSUE TITLE
Access to rejected HTTP responses

### ISSUE THREAD

**Issue by nataliagrybovska** [spray/spray#1101]
Hello,

I need a way to retrieve the status code and message of rejected HTTP response.
Let's say I want to log incoming requests and outgoing responses:

```
def logRoute: Directive0 = mapRequestContext { ctx =>
		val path = ctx.request.uri.path
		val startTime = System.currentTimeMillis()
		val id = math.abs(Random.nextLong)
		ctx.withRequestMapped { req =>
			val data = req.entity.asString
			log.info(s"Request [id]: $id, [path]: $path, [method]: ${req.method}, [data]: $data")
			req
		}.withHttpResponseMapped { resp =>
			val time = System.currentTimeMillis() - startTime
			val data = resp.entity.asString
			if (resp.status.isFailure)
				log.error(s"Error Response [id]: $id, [path]: $path, [code]: ${resp.status}, [time]: $time ms, [data]: $data")
			else
				log.info(s"Response [id]: $id, [path]: $path, [code]: ${resp.status}, [time]: $time ms, [data]: $data")
			resp
		}
	}
```

If the request is rejected there is only a way to deal with the list of rejections which cannot provide the information about HTTP status code yet. So `withRejectionsMapped` and `withRouteResponseMapped` don't help.
Is there a way to access such information at the very end when the route has been completed?

Thanks,
Natalia

**Comment by wjur** [comment]
> Is there a way to access such information at the very end when the route has been completed?

When a request is rejected, the rejection goes to RejectionHandler. Then the handler generates a status code and a response. (See: https://github.com/spray/spray/blob/master/spray-routing/src/main/scala/spray/routing/RejectionHandler.scala). There is no status code before the rejection is handled. So, I think the answer is "no".

On the other hand, you could customize the handler to log both the status code and the message (this is one of its purpose, I guess).
http://spray.io/documentation/1.1.3/spray-routing/key-concepts/rejections/#rejectionhandler

**Comment by nataliagrybovska** [comment]
Customizing the `RejectionHandler` was the last resort as there is no way to somehow override the default one. 
Seems like the only option is to copy the default rejection handler and put logging before route completion, which is extremely inconvenient.

Thanks for the help,
Natalia

**Comment by jrudolph** [comment]
@wjur is correct about how it works, but it should still be possible to achieve what you want.

What spray does is "sealing" the route passed to `runRoute`. That means it applies a default RejectionHandler (and a default ExceptionHandler), so that rejections and exceptions never escape the actor. This default RejectionHandler will handle all rejections that you haven't handled so far and, thus, convert them to status codes.

If you want to see the status codes you must make sure that you seal the route already *inside* your logging directive.

Basically you need to replicate what you see inside here:

https://github.com/spray/spray/blob/master/spray-routing/src/main/scala/spray/routing/HttpService.scala#L34

I haven't tried but it could be enough to wrap your route with sealRoute

```scala
logRoute {
  sealRoute {
    yourMainRoute
  }
}
```

**Comment by jrudolph** [comment]
Or, you can make `sealRoute` a directive like this:

```scala
def seal: Directive0 = new Directive0 {
  def happly(f: (HNil) ⇒ Route): Route = sealRoute(f(HNil))
}
```

and then include its functionality directly into your directive with

```scala
def logRoute: Directive0 = mapRequestContext { ctx => 
 //...
} & seal
```

**Comment by nataliagrybovska** [comment]
Fantastic, it works.
Thanks for the help.
